### PR TITLE
Support zero-padded RGB(a) values in Quick View

### PR DESCRIPTION
**Pull request by le717**
For #10699. This took longer than I wanted because I had to 1. track down fix in the behemoth regex and 2. update all the other tests with the new line numbers. :(

### CODE REVIEW

**Comment by MarcelGerber** [comment]
There are still multiple issues with this PR:
* All 3 arguments can start with zeros (`rgb(001, 002, 003)` is valid)
* There can be even more than 3 zeros (even though that doesn't make much sense) (`rgb(0000, 0001, 0002)` is valid)
* The same applies for `rgba()`

I'd suggest this regex (only part of it; notice the `0*`):
``` JS
/\brgb\(\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*\)/
```

**Comment by MarcelGerber** [comment]
Also, instead of manually editing all the test cases again, this may be the time to switch to [`SpecRunnerUtils.parseOffsetsFromText`](https://github.com/adobe/brackets/blob/master/test/spec/SpecRunnerUtils.js#L620) ([see it in action](https://github.com/adobe/brackets/blob/57dcc1a43b5646907a4948146591fdc8c7b6fda6/src/extensions/default/WebPlatformDocs/unittests.js#L52)).

**Comment by le717** [comment]
> There can be even more than 3 zeros (even though that doesn't make much sense) (rgb(0000, 0001, 0002) is valid)

Is that in the spec, or is it just allowed by browsers, because that is just :confused:

**Comment by redmunds** [comment]
> Is that in the spec?

Yes. In theory, there can be any number of leading zeros. See the parsing diagram for numbers: http://www.w3.org/TR/css3-syntax/#number-token-diagram

**Comment by le717** [comment]
@redmunds @MarcelGerber Regex updated and tests added. Ready for review when ready.

File: src/extensions/default/QuickView/unittest-files/test.css [hunk 1]
```diff
@@ -190,3 +190,12 @@ background: -ms-linear-gradient(top,  #d2dfed 0%,#c8d7eb 26%,#bed0ea 51%,#a6c0e3
...
     background-image: -webkit-gradient(linear, left top, left bottom, from(rgb(51,51,51)), to(rgb(204,204,204));
 }
+
+.zero_padded_rgb_rgba_values {
+    color: rgb(255, 176, 008);
+    color: rgb(001, 002, 003);
+    color: rgb(0000, 0001, 0002);
+    color: rgba(255, 176, 008, 1);
+    color: rgba(001, 002, 003, 1);
+    color: rgba(0000, 0001, 0002, 1);
```
Comment (redmunds) [inline src/extensions/default/QuickView/unittest-files/test.css]: The alpha value also supports leading zeros

File: src/utils/ColorUtils.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ define(function (require, exports, module) {
...
+    var COLOR_REGEX = new RegExp(/#[a-f0-9]{6}\b|#[a-f0-9]{3}\b|\brgb\(\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,3}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*\)|\brgb\(\s*(?:[0-9]{1,2}%|100%)\s*,\s*(?:[0-9]{1,2}%|100%)\s*,\s*(?:[0-9]{1,2}%|100%)\s*\)|\brgba\(\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,3}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*0*(?:[0-9]{1,3}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*,\s*(?:1|1\.0|0|0?\.[0-9]{1,3})\s*\)|\brgba\(\s*(?:[0-9]{1,2}%|100%)\s*,\s*(?:[0-9]{1,2}%|100%)\s*,\s*(?:[0-9]{1,2}%|100%)\s*,\s*(?:1|1\.0|0|0?\.[0-9]{1,3})\s*\)|\bhsl\(\s*(?:[0-9]{1,3})\b\s*,\s*(?:[0-9]{1,2}|100)\b%\s*,\s*(?:[0-9]{1,2}|100)\b%\s*\)|\bhsla\(\s*(?:[0-9]{1,3})\b\s*,\s*(?:[0-9]{1,2}|100)\b%\s*,\s*(?:[0-9]{1,2}|100)\b%\s*,\s*(?:1|1\.0|0|0?\.[0-9]{1,3})\s*\)|\b/.source + COLOR_NAMES.join("\\b|\\b") + "\\b", "gi");
```
Comment (redmunds) [inline src/utils/ColorUtils.js]: There can also be leading zeros for params in `hsl()` and `hsla()` function calls.

**Comment by redmunds** [comment]
@le717 This looks good, but you might as well add support for `hsl()` and `hsla()` at the same time.

File: src/utils/ColorUtils.js (hunk 2 above)
Comment (redmunds) [inline src/utils/ColorUtils.js]: There can also be leading zeros for all `%` values.

**Comment by le717** [comment]
@redmunds Might I be allowed to break up this regex into more individualized regexs (i.e. hex colors, rgb(a) colors hsl(a) colors), then concatenate them into `COLOR_REGEX`? It would make it much easier to perform changes, as right now the regex is very large and hard to read though, much less edit.

**Comment by MarcelGerber** [comment]
I think that's a good idea (but you should still only export the one we export now).
Also, replacing `[0-9]` with `\d` might also clear it up a little.

**Comment by MarcelGerber** [comment]
It's also a good idea to split each regex up into commonly reused parts, like `var 0to255 = /\s*0*(?:[0-9]{1,2}|1[0-9]{2}|2[0-4][0-9]|25[0-5])\b\s*/.source` (another one is `0to100percent`) to make any further maintenance easier.
I recommend you to use the `.source` kind-of-hack as this avoids having to care of doubled backslashs.

**Comment by redmunds** [comment]
@le717 

> Might I be allowed to break up this regex into more individualized regexs

I was thinking the same thing, but didn't mention it because there's no mechanism in regex to allow it. You'd have to build up final regex with JS (like @MarcelGerber did with color names), but I think it would make it much easier to maintain. That would be awesome.

**Comment by MarcelGerber** [comment]
@redmunds Actually, syntax like this is still quite okay imho:
```JS
var abc = /[abc]/.source,
    def = /[def].source/;
var regex = new RegExp(abc + def, "gi");
```

**Comment by redmunds** [comment]
@MarcelGerber Yes, I think that's acceptable. That's what I was referring to as "build up final regex with JS".

**Pull request closed without merging**